Sources/BraintreeCoreSwift/BTHTTP.swift

```diff
@@ -0,0 +1,542 @@
+        } else if let url = response.url, url.scheme == "data" {
+            return HTTPURLResponse(
+                url: url,
+                statusCode: 200,
+                httpVersion: nil,
+                headerFields: nil
+            )
+        }
+        return nil
+    }
+
+    func handleHTTPResponseError(response: HTTPURLResponse, data: Data, completion: (BTJSON, Error) -> Void) {
+        let responseContentType: String? = response.mimeType
+        var errorUserInfo: [String : Any] = [BTHTTPError.urlResponseKey: response]
+
+        errorUserInfo[NSLocalizedFailureReasonErrorKey] = [HTTPURLResponse.localizedString(forStatusCode: response.statusCode)]
+
+        var json: BTJSON = BTJSON()
+        if responseContentType == "application/json" {
+            json = data.isEmpty ? BTJSON() : BTJSON(data: data)
+
+            if !json.isError {
+                errorUserInfo[BTHTTPError.jsonResponseBodyKey] = json
+                let errorResponseMessage: String? = json["error"]["developer_message"].asString() ?? json["error"]["message"].asString()
+
+                if errorResponseMessage != nil {
+                    errorUserInfo[NSLocalizedDescriptionKey] = errorResponseMessage
+                }
+            }
+        }
+
+        var errorCode: BTHTTPErrorCode = response.statusCode >= 500 ? BTHTTPErrorCode.serverError : BTHTTPErrorCode.clientError
+
+        if response.statusCode == 429 {
+            errorCode = BTHTTPErrorCode.rateLimitError
+            errorUserInfo[NSLocalizedDescriptionKey] = "You are being rate-limited."
+            errorUserInfo[NSLocalizedRecoverySuggestionErrorKey] = "Please try again in a few minutes."
+        } else if response.statusCode >= 500 {
+            errorUserInfo[NSLocalizedRecoverySuggestionErrorKey] = "Please try again later."
+        }
+
+        let error = Self.constructError(code: errorCode, userInfo: errorUserInfo)
+
+        completion(json, error)
+    }
+
+    func handleJSONResponseError(
+        json: BTJSON,
+        response: URLResponse,
+        completion: @escaping (Error?) -> Void
+    ) {
+        let responseContentType: String? = response.mimeType
+        var errorUserInfo: [String : Any] = [BTHTTPError.urlResponseKey: response]
+
+        if let contentType = responseContentType, contentType != "application/json" {
+            // Return error for unsupported response type
+            errorUserInfo[NSLocalizedFailureReasonErrorKey] = "BTHTTP only supports application/json responses, received Content-Type: \(contentType)"
+            let returnedError: NSError = Self.constructError(code: .responseContentTypeNotAcceptable, userInfo: errorUserInfo)
+
+            completion(returnedError)
+        } else {
+            completion(json.asError())
+        }
+    }
+
+    static func constructError(code: BTHTTPErrorCode, userInfo: [String: Any]) -> NSError {
+        NSError(domain: BTHTTPError.domain, code: code.rawValue, userInfo: userInfo)
+    }
```

Comment:
I think we want to remove this in favor of using our new error creation pattern in swift. For example [BTDataCollectorError.swift](https://github.com/braintree/braintree_ios/blob/next-major-version/Sources/BraintreeDataCollector/BTDataCollectorError.swift).

Reply:
This one is a bit trickier as we use `domain`, `urlResponseKey`,  and `jsonResponseBodyKey` and the `BTHTTPErrorCode` enum in a bunch of the other modules. It seems like converting this and extracting those other uses out could be a larger piece of work since we use both BTHTTPError and BTHTTPErrorCode in `Card`, `PaymentFlow`, `PayPal`, `UnionPay`, and `ThreeDSecure` modules.

I think this will require a larger discussion on how we want to handle shared constants/errors across modules as a number of errors in BTHTTP will require an associated type to pass through the use info. Are we okay creating a follow up ticket for this since this class + enum are used across multiple modules?

Reply:
I think making a follow-up ticket makes sense. Could be something to do after all of the other modules have been converted.

Reply:
Agreed - a follow up ticket makes sense to me. Do we have one made already?

Reply:
Yep! Already created one for this (ends in 1898).